src/services/blocks/trace/Trace.ts

```diff
@@ -0,0 +1,690 @@
+			?.reverse()
+			.join('');
+		const hex = `0x${scale || ''}`;
+
+		return this.registry.createType('u32', hex).toBn();
+	}
+
+	/**
+	 * Parse events by
+	 * 1) Adding parent span info, event index and the storage item name to each event.
+	 * Also adds the extrinsic index as an integer if there is an `:extrinsic_index` event.
+	 * 2) Create a map of span id => array of children events
+	 * 3) Create a map of span id => to extrinisic index
+	 *
+	 * @param spansById map of span id => to span with its children for all spans
+	 * from tracing the block.
+	 */
+	private parseEvents(
+		spansById: Map<number, SpanWithChildren>
+	): {
+		eventsByParentId: EventsByParentId;
+		extrinsicIndexBySpanId: ExtrinsicIndexBySpanId;
+	} {
+		const extrinsicIndexBySpanId = new Map<number, BN>();
+		const eventsByParentId = new Map<number, ParsedEvent[]>();
+
+		for (const [idx, event] of this.traceBlock.events.entries()) {
+			const { key } = event.data.stringValues;
+			const keyPrefix = key?.slice(0, 64);
+			const storagePath = this.getStoragePathFromKey(keyPrefix);
+
+			const p = spansById.get(event.parentId);
+			const parentSpanId = p && {
+				name: p.name,
+				target: p.target,
+				id: p.id,
+			};
+
+			const parsed = { ...event, storagePath, parentSpanId, eventIndex: idx };
+
+			if (!eventsByParentId.has(parsed.parentId)) {
+				eventsByParentId.set(parsed.parentId, [parsed]);
+			} else {
+				eventsByParentId.get(parsed.parentId)?.push(parsed);
+			}
+
+			const extrinsicIndexOpt = this.maybeExtractIndex(parsed);
+			if (isSome(extrinsicIndexOpt)) {
+				// Its ok to overwrite here, we just want thee last `:extrinsic_index` `Get`
+				// in this span.
+				extrinsicIndexBySpanId.set(parsed.parentId, extrinsicIndexOpt);
+			}
+		}
+
+		return { eventsByParentId, extrinsicIndexBySpanId };
+	}
+
+	/**
+	 * Create a mapping address => Account events for that address based on an
+	 * array of all the events.
+	 *
+	 * This mapping is useful because we create operations based on consecutive
+	 * gets/puts to a single accounts balance data. So later on we can just take all
+	 * the account events from a single address and create `Operation`s from those.
+	 *
+	 * Note: Assumes passed in events are already sorted.
+	 *
+	 * @param events events with phase info
+	 */
+	accountEventsByAddress(events: ParsedActionEvent[]): AccountEventsByAddress {
+		return events.reduce((acc, cur) => {
+			if (
+				!(
+					(cur.storagePath as PalletKeyInfo).module == 'system' &&
+					(cur.storagePath as PalletKeyInfo).item == 'account'
+				)
+			) {
+				// filter out non system::Account events
+				return acc;
+			}
+
+			const asAccount = this.toAccountEvent(cur);
+			const address = asAccount.address.toString();
+
+			if (!acc.has(address)) {
+				acc.set(address, [asAccount]);
+			} else {
+				acc.get(address)?.push(asAccount);
+			}
+
+			return acc;
+		}, new Map<string, ParsedAccountEvent[]>());
```

Comment:
wouldn't Map<eventIndex, (string, ParsedAccountEvent[])> make more sense here?

Ideally you would insert that in a collection that's sorted by keys such as an MinHeap or something.

Reply:
Hmm so basically there is only ever one event with a given index, so this would just map to one event. 

I do think there is probably some room for complexity improvement in regards to sorting events. They start out sorted, but then when we create the action groups, we annotate the events with phase info and essentially create a new collection. 
```javascript
			const primarySpanEvents = eventsByParentId.get(primary.id);
			const events = primarySpanEvents?.concat(secondarySpansEvents) || [];

			events.forEach((e) => {
				actionEvents.push({
					...e,
					phase: {
						variant: phase,
						extrinsicIndex: extrinsicIndex,
					},
					primarySpanId: {
						id: primary.id,
						name: primary.name,
						target: primary.target,
					},
				});
			});
```

I suppose we could though just try out mutating the original events in place... will need to try that out to see if it makes ses

Reply:
So I think this turned into a significant improvement. We removed nlogn complexity by removing the sort we used to have to do right before creating Operations. I get around this by just always mutating the original event object. Since the events start out ordered, we just have to not mess up the ordering when creating the array of events for each address in `Map<string, ParsedAccountEvent[]>()`.

I do think the mutations may make future debugging more difficult, but seems to reduce reasonably avoidable time & space complexity.